### PR TITLE
libp2p for windows

### PR DESCRIPTION
**Pull request by lrahmani**
## Proposed changes

Add libp2p support for Windows platforms

- Add Pipe abstract class/interface and platform-dependent factory utility (for python and go)
- Add fifos-based implementation (for python and go)
- Add tcp socket-based implementation
- Refactor libp2p connection class to use abstract Pipe class
- Remove skip_test_windows mark from libp2p tests

# Fixes:

- #1422 
- #1204

### CODE REVIEW

File: packages/fetchai/connections/p2p_libp2p/aea/pipe_windows.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	buf = make([]byte, size)
+	_, err = sock.conn.Read(buf)
+	return buf, err
+}
+
+func (sock *TCPSocketPipe) Write(data []byte) error {
+	// TOFIX(LR) duplicated code to avoid circular dep
+	//    		 utils.WriteBytesConn(sock.conn, data)
+	size := uint32(len(data))
+	buf := make([]byte, 4, 4+size)
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/aea/pipe_windows.go:67]: Why here `buf := make([]byte, 4, 4+size)` and for UnixPipe Write: `buf := make([]byte, 4)`?

File: aea/helpers/pipe.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    def out_path(self) -> str:
+        return str(self._port)
+
+
+class PosixNamedPipe(LocalPortablePipe):
+    """
+    Interprocess communication implementation using Posix named pipes
+    """
+
+    def __init__(self, loop: Optional[AbstractEventLoop] = None):
```
Comment (DavidMinarsch) [inline aea/helpers/pipe.py:168]: if you add an optional logger keyword argument in constructor we can pass the logger from the connection

File: aea/helpers/pipe.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+        Returns the rendezvous point for outgoing communication
+        """
+
+
+class TCPSocketPipe(LocalPortablePipe):
+    """
+    Interprocess communication implementation using tcp sockets
+    """
+
+    def __init__(self, loop: Optional[AbstractEventLoop] = None):
```
Comment (DavidMinarsch) [inline aea/helpers/pipe.py:84]: see below re logger

File: tests/test_packages/test_connections/test_p2p_libp2p/test_public_dht.py [hunk 4]
```diff
@@ -49,8 +48,7 @@
 AEA_LIBP2P_LAUNCH_TIMEOUT = 660  # may downloads up to ~66Mb
 
 
-@skip_test_windows
-@pytest.mark.integration
+# @pytest.mark.integration
```
Comment (DavidMinarsch) [inline tests/test_packages/test_connections/test_p2p_libp2p/test_public_dht.py:51]: don't forget to reinstate

File: packages/fetchai/connections/p2p_libp2p/aea/pipe_windows.go (hunk 1 above)
Comment (lrahmani) [reply packages/fetchai/connections/p2p_libp2p/aea/pipe_windows.go:67]: well spotted, UnixPipe should be one write only as well

File: aea/helpers/pipe.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    async def connect(self, timeout=PIPE_CONN_TIMEOUT) -> bool:
+        """
+        Setup the communication channel with the other process
+        """
+
+    @abstractmethod
+    async def write(self, data: bytes) -> None:
+        """
+        Write `data` bytes to the other end of the channel
+        Will first write the size than the actual data
```
Comment (DavidMinarsch) [inline aea/helpers/pipe.py:54]: missing `param`

File: aea/helpers/pipe.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+
+class LocalPortablePipe(ABC):
+    """
+    Multi-platform interprocess communication channel
+    """
+
+    @abstractmethod
+    async def connect(self, timeout=PIPE_CONN_TIMEOUT) -> bool:
+        """
+        Setup the communication channel with the other process
```
Comment (DavidMinarsch) [inline aea/helpers/pipe.py:47]: missing param

File: aea/helpers/pipe.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+class TCPSocketPipe(LocalPortablePipe):
+    """
+    Interprocess communication implementation using tcp sockets
+    """
+
+    def __init__(
+        self,
+        logger: logging.Logger = _default_logger,
+        loop: Optional[AbstractEventLoop] = None,
+    ):
```
Comment (DavidMinarsch) [inline aea/helpers/pipe.py:94]: missing docstring

File: aea/helpers/pipe.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+        self._writer = None  # type: Optional[asyncio.StreamWriter]
+        self._loop = loop if loop is not None else asyncio.get_event_loop()
+
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        s.bind(("", 0))
+        s.listen(1)
+        self._port = s.getsockname()[1]
+        s.close()
+
+    async def connect(self, timeout=PIPE_CONN_TIMEOUT) -> bool:
```
Comment (DavidMinarsch) [inline aea/helpers/pipe.py:109]: missing docstring

**Pull request merged**